### DISCUSSION TITLE
[Bug report] An error occurred during rendering process: "undefined" is not valid JSON

### DISCUSSION THREAD

**Discussion by burnmyshadow** [GetPublii/Publii#1419] [Bug Report] [answered]
### Operating system

Windows 10

### Publii version

0.45.1 (build 16608)

### Issue type

Application

### Bug description

Hi, I can't sync my websites after upgrading to version 0.45.1 (build 16608)

```
Rendering process failed
An error occurred during rendering process:

"undefined" is not valid JSON
```

Upgraded from 0.44.x as far as I remember. Any help?

**Comment by dziudek** [comment]
Hi,

Do you have anything under Tools -> Log Viewer for rendering-errors.log and rendering-process.log?

**Reply by burnmyshadow** [reply]
rendering-errors.log — empty

rendering-process.log:

```
FEATURED TAG IMAGES - QUERY: 52.896ms
FEATURED TAG IMAGES - STORE: 161.728ms
TAGS POST COUNT - QUERY: 36.25ms
TAGS POST COUNT - STORE: 0.074ms
TAGS - QUERY: 3.761ms
TAGS - STORE: 25.432ms
FEATURED AUTHOR IMAGES - QUERY: 2.359ms
FEATURED AUTHOR IMAGES - STORE: 0.051ms
AUTHORS POST COUNT - QUERY: 5.684ms
AUTHORS POST COUNT - STORE: 0.035ms
AUTHORS - QUERY: 1.855ms
AUTHORS - STORE: 1.179ms
FEATURED POST IMAGES - QUERY: 11.573ms
FEATURED POST IMAGES - STORE: 1.862s
POST TAGS - QUERY: 7.424ms
POST TAGS - STORE: 0.843ms
POSTS - QUERY: 43.074ms
POSTS - STORE: 675.232ms
CONTENT DATA: 2.989s
COMMON DATA: 16.171ms
PARTIALS GENERATION: 17.612ms
CSS stats: 0.19592421641977575 (48255/60013)
CSS: 179.865ms
[DIFF REMOVE CATALOG] C:\Users\xxxx\Documents\Publii\sites\xxxx\preview\media\posts\292
```
(changed last line with xxxx)

**Reply by dziudek** [reply]
@burnmyshadow - not so helpful :/ Can you share with me a backup of your website so I could try to reproduce your issue with my local instance? If yes, you can send it to dziudek@gmail.com

**Reply by burnmyshadow** [reply]
Backup is too big (~2,7 Gb), maybe I can provide something else? I'll try to find a way to upload it somewhere

**Reply by burnmyshadow** [reply]
I have sent you a full backup of my site, maybe you can look at it and figure something out

**Comment by internettips** [comment]
@burnmyshadow @dziudek 

Just for info / reference, I was getting exactly the same issue on one website after upgrading to Publii 0.45.1 (build 16608) and displays the same error message when Previewing the local site.  The rendering-process -related logs don't seem to show any problems (or are empty). I'm using Apple macOS operating system, latest Sonoma version at the time of writing. The site uses the latest Publii Art Gallery theme. Let's call this Site1.

However, if I create a new 'clean' test site using the same version of Art Gallery theme, and then Preview the site, there are no errors: it all works as it should. Let's call this Site2. So ... yes, a bit weird. If a new, 'clean' website doesn't display this problem does that suggest the source problem is not a bug? Maybe the latest Publii version is more stringent in its checks compared to previous versions?

I haven't yet had time to check out the Site1 website above, but I will asap. Site1 above does have theme overrides, Site2 does not have any theme overrides. So maybe that's a clue. I'll get back if I find anything new -- if the Publii dev wizards or anyone else hasn't already come up with a fix.

**Reply by burnmyshadow** [reply]
Hi, thanks, indeed, there are theme overrides. I was able to sync a site without them too

**Reply by dziudek** [reply]
@burnmyshadow - could you share your theme overrides then ? :)

**Reply by burnmyshadow** [reply]
File was sent

**Reply by dziudek** [reply]
For me it is working so:
- it must be something related to your theme/site config
- the only solution will be probably remove file by file overrides to find a problematic one

I do not see any suspicous fragment in the overrided files which could cause such issue at this moment

**Reply by dziudek** [reply]
Also your theme uses for override Mercury v.2.2.0.0 while the newest version is v.2.6.0.0 - it also can be a reason of your issues

**Reply by burnmyshadow** [reply]
I have tried updating theme, getting the same error message. Reverting to previous version fixes the problem, but that's not optimal. Don't know what to do next really. Anyway thanks for trying helping me

**Reply by dziudek** [reply]
As I wrote earlier - you can try to remove file after file from your override (probably it will be issue with one of the .hbs files) and then check which file causes this issue

**Reply by dziudek** [reply]
Also if you have updated the theme and you are still using your override then it means that problematic file is in your override not in the updated theme.

**Reply by burnmyshadow** [reply]
I removed overrides completely, still getting error

**Reply by dziudek** [reply]
Sorry - you earlier wrote 

> I was able to sync a site without them too

So I thought that it is an issue with overrides

Did you tried to resave your theme settings before sync?

**Reply by burnmyshadow** [reply]
I have another site which I wasn't touched with overrides, it syncs ok, so I thought that was the problem. There is another theme set also

Downloaded new version of Mercury, overwritten all the files in the folder − error, ok, installed it properly via Publii (I really forgot you need to do it that way), well nothing changed. Also, version is stuck on 2.2.0.0 in Settings. Deleted every 2.2.0.0 version via interface (for some reason there were like 4 or 5 of them). Selected 2.6.0.0 it is still shows up as 2.2.0.0. Cliked Save and Preview, same error

**Reply by dziudek** [reply]
If you still see v.2.2.0.0 then it can suggest that for some reason your override is still loaded. When I have removed your override version is displayed as v.2.6.0.0

**Reply by burnmyshadow** [reply]
Okay, I figured out that I was removing overrides from wrong location (old location of old version of Publii). Now theme version shows as 2.6.0.0 but still fails to sync with overrides removed

**Reply by dziudek** [reply]
@burnmyshadow - unfortunately it is something strange - for me it is working fine - no issues. I have no idea what can cause this strange error. We did not noticed such error during our tests even with older websites. At this moment I am unable to help, because I cannot reproduce this problem on my side

**Reply by burnmyshadow** [reply]
Thank you for trying! I'll try clean install then

**Reply by burnmyshadow** [reply]
I did clean install, trying to generate preview, now I get new error
![Screenshot_2](https://github.com/GetPublii/Publii/assets/16764894/fab63745-f5e2-4f33-82c2-d7abf4de8f7a)

Publii window stuck with this progress thing
![Screenshot_3](https://github.com/GetPublii/Publii/assets/16764894/3ccd9f01-7351-4b5e-9679-7d77fbd95f9f)
It stays like that until I restart program

If I provide access data and try to sync website, progress window appears but nothing is happening
![Screenshot_4](https://github.com/GetPublii/Publii/assets/16764894/e2f053eb-767c-484c-9927-ecbfc9517958)

When I press Cancel, this error pops up
![Screenshot_5](https://github.com/GetPublii/Publii/assets/16764894/14e4b291-7c20-4da8-adb3-e2f1dd6a7c69)

Maybe I need to update something on my system

**Comment by burnmyshadow** [comment]
Checking logs, this is what I found new:

rendering-errors.log

```
node:internal/modules/cjs/loader:1084
  const err = new Error(message);
              ^

Error: Cannot find module 'C:\Users\Professional\Documents\Publii\plugins\featherIcons\main.js'
Require stack:
- C:\Users\Professional\AppData\Local\Programs\Publii\resources\app.asar\back-end\modules\render-html\renderer.js
- C:\Users\Professional\AppData\Local\Programs\Publii\resources\app.asar\back-end\workers\renderer\preview.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1084:15)
    at Module._load (node:internal/modules/cjs/loader:929:27)
    at c._load (node:electron/js2c/node_init:2:13672)
    at Module.require (node:internal/modules/cjs/loader:1150:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Renderer.loadPlugins (C:\Users\Professional\AppData\Local\Programs\Publii\resources\app.asar\back-end\modules\render-html\renderer.js:1931:34)
    at new Renderer (C:\Users\Professional\AppData\Local\Programs\Publii\resources\app.asar\back-end\modules\render-html\renderer.js:93:14)
    at process.<anonymous> (C:\Users\Professional\AppData\Local\Programs\Publii\resources\app.asar\back-end\workers\renderer\preview.js:13:24)
    at process.emit (node:events:517:28)
    at emit (node:internal/child_process:944:14) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\Professional\\AppData\\Local\\Programs\\Publii\\resources\\app.asar\\back-end\\modules\\render-html\\renderer.js',
    'C:\\Users\\Professional\\AppData\\Local\\Programs\\Publii\\resources\\app.asar\\back-end\\workers\\renderer\\preview.js'
  ]
}

Node.js v18.18.2
```

**Comment by burnmyshadow** [comment]
Ok, downloaded Feather Icons plugin. Generating preview now throws this error

![Screenshot_6](https://github.com/GetPublii/Publii/assets/16764894/71ddc7c0-0f15-4c18-916e-1347134167dc)

Enabling plugin: same error

**Reply by burnmyshadow** [reply]
Unistalling Feather Icons plugin doesn't reverts Publii to earlier errors, now all I get is the same error as above. 

rendering-errors.log is now empty

Now what should I do…

**Reply by burnmyshadow** [reply]
I should note that my site used this plugin, I played with it by inserting one shortcode somehere (still trying to find where). After clean install, there's no plugins being restored, so maybe restored backup of site references to that plugin, Publii can't find it and can't start rendering, I don't know, just my guess. Maybe something is wrong with that in this version of Publii

**Comment by clickclack69** [comment]
Same issue for me after upgrading, but not for all my websites

**Reply by burnmyshadow** [reply]
Did you install Feather Icons plugin? Maybe it is broken somehow

**Comment by clickclack69** [comment]
I haven't install this plugin.

**Comment by internettips** [comment]
Hi all,

Have finally resolved the problem on this one Publii site that originally had this error message and which stopped the website from rendering in Publii.

I temporarily replaced the existing db.sqlite with a known good / working test site db.sqlite that used the same theme, and then chose 'Preview your changes' button to preview the site. This worked: the site rendered normally. So that was my first clue.

Next after reversing the db.sqlite change, I opened the Publii site and started looking at the content. After a while, I deleted all the draft pages and tested 'Preview ...' again. This worked: the site generation worked flawlessly.

So in my instance, the problem seemed to be with one or more draft pages in Publii. After some more checking I found that if I deleted a specific 62 draft pages, 'Preview ...' and therefore rendering worked flawlessly, and of course the error message disappeared.

Another test proved that the solution wasn't anything to do with just deleting any 62 draft pages: I had to delete specific pages. It turns out those were 3 groups of pages, each group had a kind of main index page and on each of their sub-pages I would just insert a 'Forward' and 'Back' links, that just provided a quick way for users to cycle through the sequence. 

All the changes were originally done in Publii. As far as I can tell, there was no extra code, javascript, etc., on these 62 pages: just all standard pages created within the previous versions of Publii. Plus, no problems with these 62 draft web pages on previous versions of Publii.

So ... yes, a bizarre one. :-)

Hope my experience helps others to resolve their problem too.

PS: when I say 'delete' above, that means move the pages to the trash and then fully empty the trash in Publii. Just sending the 62 pages to the Publii Trash bin is not enough: they have to be fully deleted / empty the Trash bin to get Publii to update the db.sqlite database.  

PPS: update: even though the local site Previews okay and navigation appears to work normally, in the rendering-process.log, lots of pages (maybe all active pages) appear to show the same errors. Example from one page shown below:

ERR SqliteError: no such column: [Title of page here - changed for privacy purposes]
    at Database.prepare (/Applications/Publii.app/Contents/Resources/app.asar/node_modules/better-sqlite3/lib/methods/wrappers.js:5:21)
    at DBUtils.get (/Applications/Publii.app/Contents/Resources/app.asar/back-end/helpers/db.utils.js:29:28)
    at RendererContextPost.loadPost (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/contexts/post.js:207:20)
    at RendererContextPost.prepareData (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/contexts/post.js:89:18)
    at RendererContextPost.setContext (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/contexts/post.js:370:14)
    at RendererContextPost.getContext (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/contexts/post.js:430:14)
    at Renderer.generatePosts (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/renderer.js:859:44)
    at Renderer.generateWWW (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/renderer.js:227:14)
    at Renderer.renderFullPreview (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/renderer.js:181:20)
    at Renderer.renderSite (/Applications/Publii.app/Contents/Resources/app.asar/back-end/modules/render-html/renderer.js:162:28) {
  code: 'SQLITE_ERROR'
}

**Comment by internettips** [comment]
When creating a new website using Publii version: 0.45.1 (build 16608), on latest Apple macOS (Sonoma 14.3.1), using the latest Art Gallery theme v1.9.0.0, although this test site of 8 Publii pages appears to render, so can be Previewed locally, and navigated normally, in the Publii Log Viewer, the rendering-errors.log shows:

marked(): mangle parameter is enabled by default, but is deprecated since version 5.0.0, and will be removed in the future. To clear this warning, install https://www.npmjs.com/package/marked-mangle, or disable by setting `{mangle: false}`.
marked(): smartypants parameter is deprecated since version 5.0.0, should not be used and will be removed in the future. Instead use https://www.npmjs.com/package/marked-smartypants.
marked(): headerIds and headerPrefix parameters enabled by default, but are deprecated since version 5.0.0, and will be removed in the future. To clear this warning, install  https://www.npmjs.com/package/marked-gfm-heading-id, or disable by setting `{headerIds: false}`.

The rendering-process.log looks okay (= no obvious problems), ditto for rendering-deployment-process.log though it appears to show 3 oddities:

(i) Non-existing link: posts (185)
(i) Non-existing link: posts (216)
(i) Non-existing link: posts (245)

**Comment by bobmitro** [comment]
I think I got it! This happens specifically when a tag that was already set as the main tag in a post is removed. Guys, have you all removed the tag/s or is this just one possible reason?

**Comment by internettips** [comment]
@bobmitro 

In my instance Bob, no tags have been removed or changed, and no other changes made other than just upgrading Publii to the latest version, currently: 0.45.1 (build 16608) from the previous version: 0.44.4 (build 16557). Upgrading Publii seemed to be enough to trigger this 'unable to render' / Preview this particular site and display the stated error message.

**Comment by fodi** [comment]
I just encountered exactly the same error while running version 0.45.1 (build 16608), there's nothing useful in any of the log files and no errors on the developer console either. My blog is based on the Editorial-2 theme, I don't use any plugins and it's a relatively small blog (~100 MB uncompressed), @dziudek please let me know if you want me to send you an archive for further investigation.

**Comment by internettips** [comment]
@bobmitro @dziudek , and others

Further information which may be helpful in finding the cause of this weird fault, when the problem happens on this Publii website that uses Art Gallery theme, when I check the rendering-process.log , as shown below, the rendering progress seems to stop, either at rendering 'content data' below, or it completes rendering 'content data', but is unable to complete the next stage of rendering, which in a normally working system would be 'common data'.

CONFIG: 25.984ms
FEATURED TAG IMAGES - QUERY: 0.884ms
FEATURED TAG IMAGES - STORE: 0.34ms
TAGS POST COUNT - QUERY: 2.862ms
TAGS POST COUNT - STORE: 0.065ms
MAINTAGS - QUERY: 0.638ms
TAGS - QUERY: 0.719ms
TAGS - STORE: 24.611ms
FEATURED AUTHOR IMAGES - QUERY: 0.127ms
FEATURED AUTHOR IMAGES - STORE: 0.027ms
AUTHORS POST COUNT - QUERY: 0.272ms
AUTHORS POST COUNT - STORE: 0.006ms
AUTHORS - QUERY: 0.068ms
AUTHORS - STORE: 1.765ms
FEATURED POST IMAGES - QUERY: 0.658ms
FEATURED POST IMAGES - STORE: 86.546ms
POST TAGS - QUERY: 1.745ms
POST TAGS - STORE: 0.287ms
POSTS - QUERY: 1.289ms
POSTS - STORE: 80.274ms
CONTENT DATA: 243.795ms

**Comment by dziudek** [comment]
@internettips - it is for 100% related to removed main tags - that's why it stops after generating content data. It is connected with this line: https://github.com/GetPublii/Publii/commit/c1969f1be862c00aa520cdfe5aad6eabb1f287e7#diff-5b4a16da348a2a2554b48cc82f66229faec305cc592e51e14dca16177ff3a0a1R231

Hot fix is in progress.

**Reply by fodi** [reply]
Amazing, thanks!

**Comment by bobmitro** [comment, marked as answer]
FYI: Publii v.0.45.2 released and ready for [download](https://getpublii.com/download/).

**Comment by clickclack69** [comment]
It works now ! Many thanks !

**Comment by burnmyshadow** [comment]
Installed new version 0.45.2, getting new error

```
Rendering process failed
An error (1002) occurred during parsing post.hbs file.

The partial disqus could not be found
```

Installed and enabled plugin Dusqus, same error persists

**Reply by dziudek** [reply]
Most probably your theme override refers to disqus.hbs in post.hbs, while your parent theme hasn't this file (this file has been removed few versions ago from all themes, so if you are using overrides based on old version and now you have updated your theme this error will appear)

**Reply by burnmyshadow** [reply]
Yes, that was exactly what's happening. Fixed it, thank you!
